src/MemoryRouter.tsx

```diff
@@ -117,8 +115,22 @@ export class MemoryRouter extends BaseRouter {
       this.locale = options.locale;
     }
 
-    this.events.emit("routeChangeComplete");
+    this.events.emit("routeChangeStart", this.asPath, {
+      shallow: options?.shallow ?? false,
+    });
+
+    // Schedule next event emit on next tick to allow for setState updates
+    return new Promise<boolean>((resolve) =>
+      setTimeout(() => {
+        this.events.emit("routeChangeComplete", this.asPath, {
+          shallow: options?.shallow ?? false,
+        });
+        resolve(true);
+      }, 0)
```

Comment:
Two thoughts ... 
First off, you can eliminate this nesting and improve the readability like this:
```suggestion
    await new Promise(setImmediate);
    this.events.emit("routeChangeComplete", this.asPath, {
      shallow: options?.shallow ?? false,
    });
    return true;
```

Second, it's possible that `await new Promise(setImmediate)` isn't necessary.  Simply calling `await null` will probably work; it will queue execution to the end of the async queue, without actually needing to wait an entire event tick.  If that works, it's probably a simpler solution.